app/lib/common/notifications/notifications.dart

```diff
@@ -380,44 +424,52 @@ Future<bool> setupPushNotifications(
     return false;
   }
 
-  // To be informed that the device's token has been updated by the operating system
-  // You should update your servers with this token
-  Push.instance.onNewToken.listen((token) {
-    // FIXME: how to identify which clients are connected to this?
-    debugPrint('Just got a new FCM registration token: $token');
-    onNewToken(client, token);
-  });
-
   final deviceId = client.deviceId().toString();
-  final SharedPreferences preferences = await sharedPrefs();
-  final prefKey = '$deviceId.rejected_notifications';
-
-  String? token = await Push.instance.token;
-  // do we already have a token, then no need to bother the user again
-  if (token == null) {
-    // check whether we were already rejected and thus shouldn't ask again
-    if (!forced && (preferences.getBool(prefKey) ?? false)) {
-      // we need to be forced to continue
-      return false;
-    }
-    // TASK: show some extra dialog here?
-    final requested = await Push.instance.requestPermission();
-    if (!requested) {
-      // we were bluntly rejected, save and don't them bother again:
-      preferences.setBool(prefKey, false);
-      return false;
-    }
-    token = await Push.instance.token;
+  if (!forced && await wasRejected(deviceId)) {
+    // If the user rejected and we aren't asked to force, don't bother them again.
+    return false;
+  }
+  // this show some extra dialog here on devices where necessary
+  final requested = await Push.instance.requestPermission(
+    badge: true,
+    alert: true, // we request loud notifications now.
+  );
+  if (!requested) {
+    // we were bluntly rejected, save and don't them bother again:
+    await setRejected(deviceId, true);
+    return false;
   }
 
+  // let's get the token
+  final token = await Push.instance.token;
+
   if (token == null) {
+    debugPrint('No token given');
     return false;
   }
 
-  return await onNewToken(client, token);
+  return await onToken(client, token);
+}
+
+Future<bool> onNewToken(String token) async {
+  debugPrint(
+    'Received the update information for the token. Updating all clients.',
+  );
+  // ignore: use_build_context_synchronously
+  final sdk = rootNavKey.currentContext!.read(sdkProvider).requireValue;
```

Comment:
This I presume is valid in all cases, especially since this is triggering on startup ? Or if not I think we would need ```valueOrNulll``` case and handle error call in next line.

Reply:
I was thinking about that, too as both currentContext and the provider could type-wise be `null`, but I don't see any case that can actually be true in practice. And even if it was this is in a callback for a stream, so we'd see a traceback in the logs but nothing else...